contracts/USM.sol

```diff
@@ -148,21 +179,149 @@ contract USM is IUSM, ERC20Permit, Delegable {
+     *
+     * @param ethIn The amount of ETH passed to mint().
+     * @return The amount of USM, and the factor by which the ethPool grew (eg, 1.2 * WAD = 1.2x).
+     */
+    function usmFromMint(uint ethIn) public override view returns (uint, uint) {
+        // Mint USM at a sliding-up USM price (ie, at a sliding-down ETH price).  **BASIC RULE:** anytime pool_eth changes by
+        // factor k, eth_price changes by factor 1/k**2.  (Earlier versions of this logic scaled price by 1/k, not 1/k**2; but
+        // that results in calls to log()/exp().)
+        uint initialUsmPrice = usmPrice(Side.Buy);
+        uint _ethPool = ethPool();
+        uint ethPoolGrowthFactor = (_ethPool + ethIn).wadDiv(_ethPool);
+        // Math: this is an integral - sum of all USM minted at a sliding-down ETH price:
+        uint usmOut = _ethPool.wadDiv(initialUsmPrice).wadMul(WAD - WAD.wadDiv(ethPoolGrowthFactor));
+        return (usmOut, ethPoolGrowthFactor);
+    }
+
+    /**
+     * @notice How much ETH a burner currently gets from burning usmIn USM, accounting for adjustments and sliding prices.
+     *
+     * @param usmIn The amount of USM passed to burn().
+     * @return The amount of ETH, and the factor by which the ethPool shrank.
+     */
+    function ethFromBurn(uint usmIn) public override view returns (uint, uint) {
+        // Burn at a sliding-up price:
+        uint initialUsmPrice = usmPrice(Side.Sell);
+        uint _ethPool = ethPool();
+        // Math: this is an integral - sum of all USM burned at a sliding price.  Follows the same mathematical invariant as
+        // above: if pool_eth *= k, eth_price *= 1/k**2.
+        uint ethOut = WAD.wadDiv(WAD.wadDiv(_ethPool) + WAD.wadDiv(usmIn.wadMul(initialUsmPrice)));
+        uint ethPoolShrinkFactor = (_ethPool - ethOut).wadDiv(_ethPool);
+        return (ethOut, ethPoolShrinkFactor);
+    }
+
+    /**
+     * @notice How much FUM a funder currently gets back for ethIn ETH, accounting for adjustments and sliding prices.
+     *
+     * @param ethIn The amount of ETH passed to fund().
+     * @return The amount of FUM, and the factor by which the ethPool grew.
+     */
+    function fumFromFund(uint ethIn) public override view returns (uint, uint) {
+        // Create FUM at a sliding-up price:
+        uint initialFumPrice = fumPrice(Side.Buy);
+        uint _ethPool = ethPool();
+        uint ethPoolGrowthFactor;
+        uint fumOut;
+        if (_ethPool == 0 ) {
+            // This is our first ETH added, so an "ethPoolGrowthFactor" makes no sense - skip sliding-prices for this first call:
+            ethPoolGrowthFactor = type(uint).max;
+            fumOut = ethIn.wadDiv(initialFumPrice);
+        } else {
+            ethPoolGrowthFactor = (_ethPool + ethIn).wadDiv(_ethPool);
+            // Math: see closely analogous comment in mint_usm() above.
+            fumOut = _ethPool.wadDiv(initialFumPrice).wadMul(WAD - WAD.wadDiv(ethPoolGrowthFactor));
```

Comment:
`ethPoolGrowthFactor` can't be less than 1, so no need for `SafeMath.sol` here either.

Reply:
I patched it anyway though because why not?  Seems best to just use `SafeMath` everywhere rather than try to analyze each case.  No?